Source/DafnyDriver/DafnyDriver.cs

```diff
@@ -304,79 +318,95 @@ public enum CommandLineArgumentsResult {
       }
     }
 
-    private static VerificationResultCache cache;
-    public static bool Boogie(ExecutionEngineOptions options, string baseName,
-      IEnumerable<Tuple<string, Bpl.Program>> boogiePrograms, string programId,
-      out Dictionary<string, PipelineStatistics> statss, out PipelineOutcome outcome) {
+    public async Task<(bool IsVerified, PipelineOutcome Outcome, IDictionary<string, PipelineStatistics> ModuleStats)>
+      Boogie(string baseName, IEnumerable<Tuple<string, Bpl.Program>> boogiePrograms, string programId) {
+
+      var concurrentModuleStats = new ConcurrentDictionary<string, PipelineStatistics>();
+      var writerManager = new ConcurrentToSequentialWriteManager(Console.Out);
+
+      var moduleTasks = boogiePrograms.Select(async program => {
+        await using var moduleWriter = writerManager.AppendWriter();
+        // ReSharper disable once AccessToDisposedClosure
+        return await Task.Run(() => BoogieOnceWithTimer(moduleWriter, baseName, programId, concurrentModuleStats, program.Item1,
+          program.Item2));
+      }).ToList();
 
-      bool isVerified = true;
-      outcome = PipelineOutcome.VerificationCompleted;
-      statss = new Dictionary<string, PipelineStatistics>();
+      await Task.WhenAll(moduleTasks);
+      var outcome = moduleTasks.Select(t => t.Result.Outcome)
+        .Aggregate(PipelineOutcome.VerificationCompleted, MergeOutcomes);
```

Comment:
Is that the right logic?  It looks like we'll wait for the slowest of all modules, rather than stopping after the first failure?